pyomo/contrib/doe/fim_doe.py

```diff
@@ -6,11 +6,84 @@
-from pyomo.contrib.sensitivity_toolbox.sens import get_dsdp
+from pyomo.contrib.sensitivity_toolbox.sens import sipopt, sensitivity_calculation, get_dsdp
 
-class DesignOfExperiments: 
+class Measurement_flatten:
+    def __init__(self, measurement_timeset, measurement_extra_index):
+        '''
+        This class stores measurements' information
+        Parameters:
+        ----------
+        measurement_timeset: a dictionary, keys are measurement names in the model, values are a list of time they are measured
+        measurement_extra_index: A dictionary, keys are the measurement names, values are a list of indexes. If no extra indexes, values are None.
+        '''
+        self.measurement_name = list(measurement_timeset.keys())
+        self.measurement_timeset = measurement_timeset
+        self.measurement_extra_index = measurement_extra_index
+        # which measurement has extra indexes
+        self.extra_measure_name = list(measurement_extra_index.keys())
+
+    def flatten_measurement(self):
+        '''Flatten measurements and their extra index (if any)
+        Flattened names consist of their original names in the model and their extra index value, connected with a '_index_'
+        For e.g., FCO2 with an extra index 19 is 'FCO2_index_19', FCO2 without extra index is still 'FCO2'
+        This function converts all measurements passed for the DOE object
+
+        Returns
+        ------
+        flatten_measure_name: a list of flattened measurement name.
+        flatten_measure_timeset: a dictionary, keys are flattened measurement name, values are lists of measurement timepoints
+        '''
+        # check if measurement variables need to be flattened
+        if self.measurement_extra_index is not None:
+            # store flattened name
+            flatten_measure_name = []
+            # store flattened measurement time point
+            flatten_measure_timeset = {}
+            # loop over name
+            for j in self.measurement_name:
+                if j in self.extra_measure_name:
+                    for ind in self.measurement_extra_index[j]:
+                        flatten_name = j +'_index_'+ str(ind)
+                        flatten_measure_name.append(flatten_name)
+                        flatten_measure_timeset[flatten_name] = self.measurement_timeset[j]
+                else:
+                    flatten_measure_name.append(j)
+                    flatten_measure_timeset[j] = self.measurement_timeset[j]
+        # if there is no extra index, the flattened measurements will not be changed
+        else:
+            flatten_measure_name = self.measurement_name.copy()
+            flatten_measure_timeset = self.measurement_timeset.copy()
+
+        self.flatten_measure_name = flatten_measure_name
+        self.flatten_measure_timeset = flatten_measure_timeset
+        return flatten_measure_name, flatten_measure_timeset
+
+    def partly_flatten_measurement(self, jac_involved):
```

Comment:
Is this method needed anymore? Also, should this be a static method? Does it use any information in self?